### PR TITLE
add question authoring guard

### PR DESCRIPTION
**Pull request by aakano684**
## Purpose

Add question authoring guard for unauthorized users
- ...

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type

What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

- Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

- Test the code

```

```

## What to Check

Verify that the following are valid

-

### CODE REVIEW

File: src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { TestBed } from '@angular/core/testing';
+
+import { QuestionAuthoringGuard } from './question-authoring.guard';
+import { AuthenticationService } from '../../../services/authentication.service';
+import { User } from '../../../models/user';
+import { UserPermission } from '../../../models/permissions';
+import { of, Observable } from 'rxjs';
+import { Router } from '@angular/router';
+
+describe('QuestionAuthoringGuard', () => {
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts:10]: Really liking the tests and use of mocking here.

File: src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    guard = TestBed.inject(QuestionAuthoringGuard);
+    authService = TestBed.inject(AuthenticationService);
+  });
+
+  it('should be created', () => {
+    expect(guard).toBeTruthy();
+  });
+
+
+  it('should allow the authenticated user to access app', () => {
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts:30]: As this is a unit test, it's probably worth repeating the scenario for the user's we would expect to be able to access this. For example, `org-admin` should be able to create questions and `creator` should be able to do likewise.

File: src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      belongsToOrgWithId: 'default'
+    });
+    spyOnProperty(authService, 'currentUser$').and.returnValue(of(user));
+    expect(guard.canActivate(routeMock, routeStateMock)).toEqual(true);
+  });
+
+  it('should not allow the unauthenticated user to access app', () => {
+    const user = new User({
+      name: 'test',
+      permissions: new Set([UserPermission.CREATOR]),
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts:43]: I would have thought a `creator` should be able to create questions. The ones I expect wouldn't have permissions for this are if a user has only `consumer` or only `global-admin`.

File: src/app/components/practice/question-authoring-page/question-authoring.guard.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  constructor(
+    private authenticationService: AuthenticationService
+) { }
+  canActivate(
+    next: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot): boolean {
+      let isPermitted  = false;
+      this.authenticationService.currentUser$.subscribe( x => x.permissions.
+           has(UserPermission.ORG_ADMIN) ? isPermitted = true : isPermitted = false);
+      return isPermitted;
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.ts:19]: I'm not sure if this variable will be set by the time the synchronous execution has happened.

Comment (noce2) [reply src/app/components/practice/question-authoring-page/question-authoring.guard.ts:19]: It turns out, https://angular.io/api/router/CanActivate, you can return an observable with a Boolean in it and it will be happy :). 

This means you can use the `.pipe` operator instead of subscribe and return the result of the Boolean operation you're doing within the operator.

**Review by noce2 — CHANGES_REQUESTED** [review]
Hey, looks good. Some smaller issues related to observable handling and unit test case coverage.

File: src/app/components/practice/question-authoring-page/question-authoring.guard.ts [hunk 5]
```diff
@@ -12,11 +14,10 @@ export class QuestionAuthoringGuard implements CanActivate {
 ) { }
   canActivate(
     next: ActivatedRouteSnapshot,
-    state: RouterStateSnapshot): boolean {
-      let isPermitted  = false;
-      this.authenticationService.currentUser$.subscribe( x => x.permissions.
-           has(UserPermission.ORG_ADMIN) ? isPermitted = true : isPermitted = false);
-      return isPermitted;
+    state: RouterStateSnapshot): Observable<boolean> {
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.ts:17]: 👍

File: src/app/components/practice/question-authoring-page/question-authoring.guard.ts [hunk 6]
```diff
@@ -12,11 +14,10 @@ export class QuestionAuthoringGuard implements CanActivate {
...
+    return this.authenticationService.currentUser$.pipe(map((user) => {
+        return ((user.permissions.
+        has(UserPermission.ORG_ADMIN) || user.permissions.has(UserPermission.CREATOR)));
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.ts:20]: 👍

File: src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts [hunk 7]
```diff
@@ -27,23 +27,51 @@ describe('QuestionAuthoringGuard', () => {
...
+      permissions: new Set([UserPermission.CONSUMER]),
+      belongsToOrgWithId: 'default'
+    });
+    spyOnProperty(authService, 'currentUser$').and.returnValue(of(user));
+    guard.canActivate(routeMock, routeStateMock).subscribe((resp) => {
+      expect(resp).toEqual(false);
+    });
+  });
+
+  it('should not allow global admin to access authoring page', () => {
```
Comment (noce2) [inline src/app/components/practice/question-authoring-page/question-authoring.guard.spec.ts:66]: 👍

**Review by noce2 — APPROVED** [review]
LGTM. Nice addition to the tests, happy to approve from here. 

One edge case you might consider including in the unit tests is, what happens if a user's permissions set contains a required and non-required one? For example, if a user is both a `consumer` and `creator`. Would that mean they got rejected? Now we both know from your implementation mechanism (you just check the required permissions are there) that that won't happen but if in the future, someone changed that, we would suddenly get unexpected behaviour. An edge case unit test like that would guard (pun intended 😄 ) against that.
